lib/vrl/stdlib/src/hmac.rs

```diff
@@ -0,0 +1,209 @@
+    };
+    Ok(hash.into())
+}
+
+#[derive(Clone, Copy, Debug)]
+pub struct Hmac;
+
+impl Function for Hmac {
+    fn identifier(&self) -> &'static str {
+        "hmac"
+    }
+
+    fn parameters(&self) -> &'static [Parameter] {
+        &[
+            Parameter {
+                keyword: "value",
+                kind: kind::BYTES,
+                required: true,
+            },
+
+            Parameter {
+                keyword: "key",
+                kind: kind::BYTES,
+                required: true,
+            },
+
+            Parameter {
+                keyword: "algorithm",
+                kind: kind::BYTES,
+                required: false
+            },
+
+            Parameter {
+                keyword: "encoding",
+                kind: kind::BYTES,
+                required: false
+            }
+        ]
+    }
+
+    fn examples(&self) -> &'static [Example] {
+        &[
+            Example {
+                title: "default (SHA-256, base64-encoded result)",
+                source: r#"hmac("Hello there", "supersecretkey")"#,
+                result: Ok("kmpc79vrb6SODvg4LwivUnb443+IhR9SSW55KcBPKo8="),
+            },
+
+            Example {
+                title: "SHA-256, hex encoded result",
+                source: r#"hmac("Hello there", "supersecretkey", encoding: "hex")"#,
+                result: Ok("926a5cefdbeb6fa48e0ef8382f08af5276f8e37f88851f52496e7929c04f2a8f")
+            },
+
+            Example {
+                title: "SHA1, base64-encoded result",
+                source: r#"hmac("Hello there", "supersecretkey", algorithm: "SHA1")"#,
+                result: Ok("795HKoopDtOb45EOUroxeHz1OWo=")
+            }
+        ]
+    }
+
+    fn compile(
+        &self,
+        _state: &state::TypeState,
+        _ctx: &mut FunctionCompileContext,
+        arguments: ArgumentList,
+    ) -> Compiled {
+        let value = arguments.required("value");
+        let key = arguments.required("key");
+        let algorithm = arguments.optional("algorithm");
+        let encoding = arguments.optional("encoding");
+
+        Ok(HmacFn { value, key, algorithm, encoding }.as_expr())
+    }
+}
+
+#[derive(Debug, Clone)]
+struct HmacFn {
+    value: Box<dyn Expression>,
+    key: Box<dyn Expression>,
+    algorithm: Option<Box<dyn Expression>>,
+    encoding: Option<Box<dyn Expression>>,
+}
+
+impl FunctionExpression for HmacFn {
+    fn resolve(&self, ctx: &mut Context) -> Resolved {
+        let value = self.value.resolve(ctx)?;
+        let key = self.key.resolve(ctx)?;
+        let algorithm = match &self.algorithm {
+            Some(algorithm) => algorithm.resolve(ctx)?,
+            None => value!("SHA-256")
+        };
+        let encoding = match &self.encoding {
+            Some(encoding) => encoding.resolve(ctx)?,
+            None => value!("base64")
+        };
+
+        hmac(value, key, algorithm, encoding)
+    }
+
+    fn type_def(&self, _: &state::TypeState) -> TypeDef {

```

Comment:
I think one of my previous comments here was missed: https://github.com/vectordotdev/vector/pull/15371#discussion_r1108976559

The fallibility can be determined based on the inputs. For example, if the inputs are known to be valid at compile-time, this can be marked as infallible.
Example infallible call:
```coffee
hmac("value", "key", "SHA-256", "encoding", "hex")
```

Example fallible call: (value of `algorithm` / `encoding` are unknown)
```coffee
hmac("value", "key", .algorithm, .encoding)
```